OpenRA.Mods.Common/FileFormats/ImaAdpcmReader.cs

```diff
@@ -85,5 +99,108 @@ public static byte[] LoadImaAdpcmSound(ReadOnlySpan<byte> raw, ref int index, re
 
 			return output;
 		}
+
+		public static byte[] ReadData(Stream stream, long dataOffset, int dataSize, short blockAlign, short channels)
+		{
+			const int SamplesPerGroup = 8;
+
+			ArgumentNullException.ThrowIfNull(stream);
+
+			var sourceData = SegmentStream.GetReadableData(stream, dataOffset, dataSize);
+
+			var numBlocks = dataSize / blockAlign;
+
+			var predictorSize = 4 * channels;
+			var blockDataSize = blockAlign - predictorSize;
+
+			// We get two samples per nibble
+			var samplesPerChannel = blockDataSize * 2 / channels;
+
+			// 8 samples from a 4-byte group
+			var groupCount = samplesPerChannel / SamplesPerGroup;
+
+			var predOut = numBlocks * channels * 2;
+			var groupOut = numBlocks * groupCount * channels * SamplesPerGroup * 2;
+			var estimatedOutDataSize = predOut + groupOut;
+
+			var outData = new byte[estimatedOutDataSize];
+
+			// PERF: The output is 16-bit PCM, so we can write bytes as if they were shorts for less CPU churn.
+			var outShorts = MemoryMarshal.Cast<byte, short>(outData.AsSpan());
+
+			// NOTE: decoding a block is sequentually dependent on predictor/index.
+			Span<short> predictor = stackalloc short[channels];
+			Span<byte> index = stackalloc byte[channels];
+
+			// global decoded sample counter
+			var outSample = 0;
+			var src = 0;
+			for (var block = 0; block < numBlocks && src < sourceData.Length; block++)
+			{
+				// Initial states
+				for (var c = 0; c < channels; c++)
+				{
+					if (src + 4 > sourceData.Length)
+						return outData;
+
+					// Load initial values.
+					predictor[c] = (short)(sourceData[src] | (sourceData[src + 1] << 8));
+					index[c] = sourceData[src + 2];
+
+					// sourceData[src + 3] is unknown/reserved
+					src += 4;
+				}
+
+				// Write initial predictor samples interleaved
+				for (var c = 0; c < channels; c++)
+					outShorts[outSample + c] = predictor[c];
+
+				outSample += channels;
+
+				for (var iter = 0; iter < groupCount; iter++)
+				{
+					if (src + 4 * channels > sourceData.Length)
+						return outData;
+
+					// Decode 8 samples sequentially per channel
+					for (var c = 0; c < channels; c++)
+					{
+						ref var pred = ref predictor[c];
+						ref var idx = ref index[c];
+
+						var b0 = sourceData[src];
+						var b1 = sourceData[src + 1];
+						var b2 = sourceData[src + 2];
+						var b3 = sourceData[src + 3];
+						src += 4;
+
+						// Decode into temporary variables so they could be easily turned into registers.
```

Comment:
what is the added value is directly after assign the values to the array?

now you may force values on the stack.